### ISSUE TITLE
:man_farmer: 5 Test regressions in nightly_osx_debug

### ISSUE THREAD

**Issue by Blast545** [ros2/rclcpp#1787]
## Bug report

CI Test regressions, this causes 5 packages to segfault before the test exits normal execution.

Reference build: https://ci.ros2.org/job/nightly_osx_debug/2158/
First build with these errors: https://ci.ros2.org/job/nightly_osx_debug/2159/

Failing tests:
```
projectroot.test_kdl
projectroot.tf2_eigen_test
projectroot.test_tf2_geometry_msgs
projectroot.test_tf2_bullet
projectroot.test_tf2_sensor_msgs_cpp
```

I'm opening the issue here because after taking a look to the differences between the repo packages between these two runs, the main culprit appears to be this PR #1748.

**Required Info:**

- Operating System: MacOS
- Installation type: Source

#### Expected behavior
Test run and exit cleanly.

#### Actual behavior
5 Tests are passing, but segfaulting when closed.

#### Additional information
I opened a CI run to test if it fails when a failing test is isolated:
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=12997)](http://ci.ros2.org/job/ci_osx/12997/)

**Comment by Blast545** [comment]
Without context I would say it's something related to trying to access the global context after shutdown?

https://github.com/ros2/rclcpp/blob/fa3a6fa597c5d40b2fce46e5fb95f541e62076e7/rclcpp/src/rclcpp/clock.cpp#L75-L79

**Comment by emersonknapp** [comment]
Ah - yeah that seems like it would be an issue. Interesting that it doesn't show up on other platforms. Perhaps a simple nullptr check on the global default context would fix it?

**Comment by Blast545** [comment]
FYI this caused a regression in aarch64 nighlies as well:

[aarch64_debug#1736](https://ci.ros2.org/view/nightly/job/nightly_linux-aarch64_debug/1736/)
[aarch64_debug#1737](https://ci.ros2.org/view/nightly/job/nightly_linux-aarch64_debug/1737/)
[aarch64_debug#1738](https://ci.ros2.org/view/nightly/job/nightly_linux-aarch64_debug/1738/)

Although the problem there seems to be that one of the tests is flaky, because today's nighlies are green.

**Comment by Blast545** [comment]
Closed with #1793
